bin/liquidator/src/rebalance.rs

```diff
@@ -222,47 +249,93 @@ impl Rebalancer {
         Ok((sig, route))
     }
 
-    async fn determine_best_jupiter_tx(
+    fn get_alternative_token_amount(
         &self,
-        full: &jupiter::Quote,
-        alternatives: &[jupiter::Quote],
-    ) -> anyhow::Result<(TransactionBuilder, jupiter::Quote)> {
-        let builder = self
-            .mango_client
-            .jupiter()
-            .prepare_swap_transaction(full)
-            .await?;
-        let tx_size = builder.transaction_size()?;
-        if tx_size.is_within_limit() {
-            return Ok((builder, full.clone()));
-        }
-        trace!(
-            route_label = full.first_route_label(),
-            %full.input_mint,
-            %full.output_mint,
-            ?tx_size,
-            limit = ?TransactionSize::limit(),
-            "full route does not fit in a tx",
-        );
+        in_token_index: &u16,
+        in_amount_quote: u64,
+    ) -> anyhow::Result<(Pubkey, u64)> {
+        let in_token: &TokenContext = self.mango_client.context.token(*in_token_index);
+        let in_price = self
+            .account_fetcher
+            .fetch_bank_price(&in_token.first_bank())?;
+        let in_amount = (I80F48::from(in_amount_quote) / in_price)
+            .ceil()
+            .to_num::<u64>();
+
+        Ok((in_token.mint, in_amount))
+    }
 
-        if alternatives.is_empty() {
-            anyhow::bail!(
-                "no alternative routes from {} to {}",
-                full.input_mint,
-                full.output_mint
+    async fn determine_best_swap_tx(
+        &self,
+        mut routes: Vec<swap::Quote>,
+        input_mint: Pubkey,
+        output_mint: Pubkey,
+    ) -> anyhow::Result<(TransactionBuilder, swap::Quote)> {
+        let mut prices = HashMap::<Pubkey, I80F48>::new();
+        let mut get_or_fetch_price = |m| {
+            prices.get(&m).copied().unwrap_or_else(|| {
+                let token = self
+                    .mango_client
+                    .context
+                    .token_by_mint(&m)
+                    .expect("token for mint not found");
+                let p = self
+                    .account_fetcher
+                    .fetch_bank_price(&token.first_bank())
+                    .expect("failed to fetch price");
+                prices.insert(m, p);
+                p
+            })
+        };
+
+        routes.sort_by_cached_key(|r| {
+            let in_price = get_or_fetch_price(r.input_mint);
+            let out_price = get_or_fetch_price(r.output_mint);
+            let amount = out_price * I80F48::from_num(r.out_amount)
+                - in_price * I80F48::from_num(r.in_amount);
```

Comment:
mmh, this is not a sufficient criterion now that you've removed the favoring of the primary route

say someone wants to swap mSOL -> KIN. Due to the configuration we also check the alternate route mSOL -> SOL. This latter route is near guaranteed to have better `amount` than the one to KIN. But it will still often do worse if you factor in the SOL -> KIN second step that's needed afterwards.

suggestion: two step approach?
- fetch the routes between the correct in and out mints first (including sanctum if enabled and it's both LSTs) and find the best route (by `amount`) that fits in a transaction
- if this fails, fetch the alternate routes and grab the best one of those the same way

that also reduces the number of quote requests sent to the swap endpoints (being rate limited is a real concern)